lldb/source/Expression/DiagnosticManager.cpp

```diff
@@ -48,8 +48,17 @@ std::string DiagnosticManager::GetString(char separator) {
   std::string ret;
 
   for (const auto &diagnostic : Diagnostics()) {
-    ret.append(StringForSeverity(diagnostic->GetSeverity()));
-    ret.append(std::string(diagnostic->GetMessage()));
+    std::string message(diagnostic->GetMessage());
+    std::string searchable_message(diagnostic->GetMessage().lower());
+    std::string severity(StringForSeverity(diagnostic->GetSeverity()));
+
+    // Erase the (first) redundant severity string in the message.
+    size_t position = searchable_message.find(severity);
+    if (position != std::string::npos)
+      message.erase(position, severity.length());
+
+    ret.append(severity);
+    ret.append(message);
```

Comment:
I'm not sure how hot this code path is, but we are doing a bunch of unnecessary string copies in here (as we were before this PR as well...). For example, `message` and `severity` are both `StringRef`s that don't need to be converted into strings.

Suggestion:

```c++
std::string DiagnosticManager::GetString(char separator) {
  std::string ret;
  llvm::raw_string_ostream stream(ret);

  for (const auto &diagnostic : Diagnostics()) {
    llvm::StringRef severity = StringForSeverity(diagnostic->GetSeverity());
    stream << severity;

    llvm::StringRef message = diagnostic->GetMessage();
    std::string searchable_message = message.lower();
    auto severity_pos = message.find_first_of(severity);
    stream << message.take_front(severity_pos);

    if (severity_pos != llvm::StringRef::npos)
      stream << message.drop_front(severity_pos + severity.size());
    stream << separator;
  }

  return ret;
}
```

Reply:
I thought `StringRef` was better for strings that get passed around and reused and `std::string` is better for transient work, such as this.

Is general, is there an advantage to using `StringRef` over `std::string`?

Reply:
I don't know how hot it is either but, I do know `UserExpression.cpp` has 2 call sites, which may be a popular path.

There's about 15 call sites from these files:
* `lldb/include/lldb/Expression/DiagnosticManager.h`
* `lldb/source/Breakpoint/BreakpointLocation.cpp`
* `lldb/source/Expression/DiagnosticManager.cpp`
* `lldb/source/Expression/UserExpression.cpp`
* `lldb/source/Expression/UtilityFunction.cpp`
* `lldb/source/Interpreter/CommandReturnObject.cpp`
* `lldb/source/Plugins/ExpressionParser/Clang/ClangExpressionParser.cpp`
* `lldb/source/Plugins/Platform/POSIX/PlatformPOSIX.cpp`
* `lldb/source/Plugins/Platform/Windows/PlatformWindows.cpp`
* `lldb/source/Target/Target.cpp`
* `lldb/source/Utility/Status.cpp`

The one from `DiagnosticManager.h` is its `Dump()` method which itself has 7 call sites.

I have a future PR that'll move this functionality into `class Status`, and whatever we decided to do here we can do there as well.

Reply:
> I thought StringRef was better for strings that get passed around and reused and std::string is better for transient work, such as this.

> Is general, is there an advantage to using StringRef over std::string?


A StringRef is a _non-owning view_ into a string. When we see one of those, that means the author of the code is making a few promises:
1. They don't need to mutate the string
2. They don't need to extend its lifetime past "the current" context.

Conversely, when the author conjures a `std::string` out of `StringRef`, the author is saying that they intend to break one of those promises. Note that your code doesn't need to, with the exception `searchable_message`: we had to modify the string with "to_lower"!

Reply:
Oh, also `ret` breaks both promises: we modify it  a lot and we also extend its lifetime past the current function

Reply:
Ok, you're suggestion is up @felipepiovezan.
Thanks again! 🙂